subprojects/logging/src/main/java/org/gradle/internal/logging/sink/GroupingProgressLogEventGenerator.java

```diff
@@ -0,0 +1,180 @@
+    private final Map<Object, Object> buildOpIdHierarchy = new LinkedHashMap<Object, Object>();
+    private final Map<Object, OperationGroup> operationsInProgress = new LinkedHashMap<Object, OperationGroup>();
+    private final Map<OperationIdentifier, Object> progressToBuildOpIdMap = new LinkedHashMap<OperationIdentifier, Object>();
+
+    private Object lastRenderedBuildOpId;
+
+    public GroupingProgressLogEventGenerator(OutputEventListener listener, TimeProvider timeProvider) {
+        this.listener = listener;
+        this.timeProvider = timeProvider;
+    }
+
+    public void onOutput(OutputEvent event) {
+        if (event instanceof ProgressStartEvent) {
+            onStart((ProgressStartEvent) event);
+        } else if (event instanceof RenderableOutputEvent) {
+            handleOutput((RenderableOutputEvent) event);
+        } else if (event instanceof ProgressCompleteEvent) {
+            onComplete((ProgressCompleteEvent) event);
+        } else if (event instanceof EndOutputEvent) {
+            onEnd((EndOutputEvent) event);
+        } else if (!(event instanceof ProgressEvent)) {
+            listener.onOutput(event);
+        }
+    }
+
+    private void onStart(ProgressStartEvent startEvent) {
+        Object buildOpId = startEvent.getBuildOperationId();
+        if (buildOpId != null) {
+            buildOpIdHierarchy.put(buildOpId, startEvent.getParentBuildOperationId());
+            progressToBuildOpIdMap.put(startEvent.getProgressOperationId(), buildOpId);
+
+            // Create a new group for tasks or configure project
+            if (isGroupedOperation(startEvent.getBuildOperationType())) {
+                String header = startEvent.getLoggingHeader() != null ? startEvent.getLoggingHeader() : startEvent.getDescription();
+                operationsInProgress.put(buildOpId, new OperationGroup(startEvent.getCategory(), header, startEvent.getTimestamp(), startEvent.getBuildOperationId()));
+            }
+        }
+    }
+
+    private boolean isGroupedOperation(BuildOperationType buildOperationType) {
+        return buildOperationType == BuildOperationType.TASK || buildOperationType == BuildOperationType.CONFIGURE_PROJECT;
+    }
+
+    private void handleOutput(RenderableOutputEvent event) {
+        Object operationId = getOperationId(event.getBuildOperationId());
+        if (operationId != null) {
+            operationsInProgress.get(operationId).bufferOutput(event);
+        } else {
+            onUngroupedOutput(event);
+        }
+    }
+
+    private void onComplete(ProgressCompleteEvent completeEvent) {
+        Object buildOpId = progressToBuildOpIdMap.get(completeEvent.getProgressOperationId());
+        if (operationsInProgress.containsKey(buildOpId)) {
+            operationsInProgress.remove(buildOpId).flushOutput();
+        }
+    }
+
+    private void onEnd(EndOutputEvent event) {
+        for (Map.Entry<Object, OperationGroup> entry : operationsInProgress.entrySet()) {
```

Comment:
I think we can just iterate over the value set. That will make the code a bit easier to read.